OOInteraction/src/commands/CAddNodeToViewByName.cpp

```diff
@@ -0,0 +1,155 @@
+ **
+ **********************************************************************************************************************/
+
+#include "CAddNodeToViewByName.h"
+
+#include "OOModel/src/declarations/Declaration.h"
+#include "VisualizationBase/src/items/ViewItem.h"
+#include "OOModel/src/declarations/Declaration.h"
+#include "OOModel/src/declarations/Class.h"
+#include "OOModel/src/declarations/Method.h"
+#include "OOModel/src/declarations/Module.h"
+#include "OOModel/src/declarations/Project.h"
+
+#include "ModelBase/src/model/TreeManager.h"
+
+
+namespace OOInteraction {
+
+CAddNodeToViewByName::CAddNodeToViewByName() : Command{"add"}{}
+
+bool CAddNodeToViewByName::canInterpret(Visualization::Item*, Visualization::Item*,
+		const QStringList& commandTokens, const std::unique_ptr<Visualization::Cursor>&)
+{
+	return commandTokens.size() >= 1 && commandTokens.first() == name();
+}
+
+Interaction::CommandResult* CAddNodeToViewByName::execute(Visualization::Item* source, Visualization::Item*,
+		const QStringList& commandTokens, const std::unique_ptr<Visualization::Cursor>&)
+{
+	if (commandTokens.size() < 2)
+		return new Interaction::CommandResult(new Interaction::CommandError("Please specify a node to add"));
+
+	if (auto node = findNode(commandTokens[1],
+			Model::AllTreeManagers::instance().loadedManagers().first()->root()))
+	{
+		source->scene()->currentViewItem()->insertNode(node);
+		return new Interaction::CommandResult();
+	}
+	else
+		return new Interaction::CommandResult(new Interaction::CommandError(
+					"Could not find node with name " + commandTokens[1]));
+}
+
+QList<Interaction::CommandSuggestion*> CAddNodeToViewByName::suggest(Visualization::Item*, Visualization::Item*,
+		const QString& textSoFar, const std::unique_ptr<Visualization::Cursor>&)
+{
+	QList<Interaction::CommandSuggestion*> suggestions;
+
+	if (textSoFar.trimmed().startsWith("add ", Qt::CaseInsensitive))
+	{
+		auto nodeName = textSoFar.trimmed().mid(4);
+		auto names = findNames(nodeName.split("_"),
+						Model::AllTreeManagers::instance().loadedManagers().first()->root());
+		//Shorter names usually have less parts to the fully qualified name -> suggest them first
+		std::sort(names.begin(), names.end(), [](QString first, QString second)
+											{ return first.length() < second.length(); });
+		//Limit the number of suggestions
+		names = names.mid(0, 10);
+		for (auto name : names)
+			suggestions.append(new Interaction::CommandSuggestion("add " + name,
+									"Add node " + name + " to the view"));
+	}
+	else if (QString("add ").startsWith(textSoFar.trimmed(), Qt::CaseInsensitive) )
+			suggestions.append(new Interaction::CommandSuggestion("add ", "Add nodes to the current view"));
+
+	return suggestions;
+}
+
+QStringList CAddNodeToViewByName::findNames(QStringList nameParts, Model::Node* root)
+{
+	auto declaration = DCast<OOModel::Declaration>(root);
+	QStringList result;
+	//If it isn't a declaration, just pass it on
+	if (!declaration)
```

Comment:
This is not the right way to search for names, mainly because it is specific to OO. You should not care at all about declarations or other things. What you should do is just traverse the tree and for each node use the `node->definesSymbol` and `node->symbolName`.